protocols/request-response/src/codec.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+pub use libp2p_core::ProtocolName;
+
+use futures::{prelude::*, future::BoxFuture};
+use std::io;
+
+/// A `RequestResponseCodec` defines the request and response types
+/// for a [`RequestResponse`](crate::RequestResponse) protocol or
+/// protocol family and how they are encoded / decoded on an I/O stream.
+pub trait RequestResponseCodec {
+    /// The type of protocol(s) or protocol versions being negotiated.
+    type Protocol: ProtocolName + Send + Sync + Clone;
+    /// The type of inbound and outbound requests.
+    type Request: Send + Clone;
```

Comment:
It is a requirement on inbound events to any `ProtocolsHandler` (due to the general possibility of `NotifyHandler::All`).

Reply:
Where in the type-system is this enforced?

Reply:
The `ExpandedSwarm` [needs `Clone` on `TInEvent`](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/lib.rs#L326) since [`notify_all` needs `Clone` on `TInEvent`](https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/lib.rs#L855). Is that what you were asking for?

Reply:
I noticed that removing `Clone` does not cause compilation to fail. Should the bound not be required by the `ProtocolsHandler::InEvent` type?

Reply:
> I noticed that removing Clone does not cause compilation to fail.

Indeed, that is true, the bound here does not seem necessary. Here `InEvent` is `struct RequestProtocol<TCodec>` which has a field of type `TCodec::Request` together with `[derive(Clone)]`. So certainly `RequestProtocol` cannot be `Clone` without `TCodec::Request` being `Clone` (and sure enough the included test with a ping/pong protocol complains firstly about the missing `Clone` for `struct Ping` due to the requirement for `TInEvent: Clone` from the `Swarm`). So the `TCodec::Request: Clone` bound is probably "implied" by the `Clone` derive.

> Should the bound not be required by the ProtocolsHandler::InEvent type?

You mean as an alternative to putting the bound in the `ExpandedSwarm` impl? That would probably work - we have many such cases where the `Swarm` or `Network` is not really useable without certain constraints on some associated types, yet the bounds are placed on methods or impls of the `Swarm` or `Network`, instead of on the associated type. I personally would prefer the bounds to be directly on the associated types in these cases, but it is probably not something for this PR.

In any case, I will remove the seemingly unnecessary `Clone` constraint here, thanks for pointing that out.